flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/util/FlinkPackage.java

```diff
@@ -19,15 +19,31 @@
 package org.apache.iceberg.flink.util;
 
 import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
 
 public class FlinkPackage {
-  /** Choose {@link DataStream} class because it is one of the core Flink API. */
-  private static final String VERSION = DataStream.class.getPackage().getImplementationVersion();
+
+  public static final String FLINK_UNKNOWN_VERSION = "Flink-UNKNOWN";
 
   private FlinkPackage() {}
 
   /** Returns Flink version string like x.y.z */
   public static String version() {
-    return VERSION;
+    try {
+      String version = getVersionFromJar();
+      /* If we can't detect the exact implementation version from the jar (this can happen if the DataStream class
+       appears multiple times in the same classpath such as with shading), then the best we can do is say it's
+       unknown
+      */
+      return version != null ? version : FLINK_UNKNOWN_VERSION;
```

Comment:
I think we should store the result in a variable rather than re-fetching the version on every call

Reply:
+1. this method can be used to initialize the static `VERSION` variable

Reply:
@nastra @stevenzwu - Iceberg's checkstyle config forbids naming a non-final static variable an all-caps name; it wants camel-case. 

Should I:
1. name it "version" OR
2. Name it VERSION and add a SuppressWarning flag OR
3. Name it VERSION and fix the checkstyle.xml to use the same regex as static constants use

Reply:
I think we might want to slightly update the implementation to 
```
public class FlinkPackage {

  private static final AtomicReference<String> VERSION = new AtomicReference<>();
  private static final String UNKNOWN_VERSION = "UNKNOWN-VERSION";

  private FlinkPackage() {}

  /** Returns Flink version string like x.y.z */
  public static String version() {
    if (null == VERSION.get()) {
      VERSION.set(initFlinkVersion());
    }

    return VERSION.get();
  }

  private static String initFlinkVersion() {
    try {
      String version = versionFromJar();
      // use unknown version in case exact implementation version can't be found from the jar (this
      // can happen if the DataStream class appears multiple times in the same classpath such as
      // with shading)
      return version != null ? version : UNKNOWN_VERSION;
    } catch (Exception e) {
      return UNKNOWN_VERSION;
    }
  }

  @VisibleForTesting
  static String versionFromJar() {
    // Choose {@link DataStream} class because it is one of the core Flink API
    return DataStream.class.getPackage().getImplementationVersion();
  }
}
```
Notice that I also removed the `get` prefix from `versionFromJar()` as Iceberg typically doesn't use `get` methods

Reply:
If we do this, we might not want to create an extra method for `initFlinkVersion`, just something like this:
```
  public static String version() {
    if (null == VERSION.get()) {
      try {
        String jarVersion = versionFromJar();
        // use unknown version in case exact implementation version can't be found from the jar (this
        // can happen if the DataStream class appears multiple times in the same classpath such as
        // with shading)
        VERSION.set(jarVersion != null ? jarVersion : UNKNOWN_VERSION);
      } catch (Exception e) {
        VERSION.set(UNKNOWN_VERSION);
      }
    }

    return VERSION.get();
  }
```

Reply:
would this simple way work, where initFlinkVersion implements the try-catch?
```
private static final String VERSION = initFlinkVersion();
```

Reply:
@stevenzwu - making VERSION `private static final String` makes it untestable, because you can't test the normal and default code paths in the same JVM. 

@nastra @pvary - Using an AtomicReference would work, and if that's what you want I'll do it, but isn't that more complexity than this logic requires? `private static volatile String` works just fine, aside from the checkstyle issue mentioned above -- I need to either tweak the checkstyle settings, or give it a camelCase name.

Reply:
Uploaded a 1-method version using `AtomicReference` similar to what @nastra and @pvary were suggesting.